src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala

```diff
@@ -0,0 +1,663 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+package com.fulcrumgenomics.fastq
+
+import java.io.File
+import java.nio.file.Path
+import java.util.NoSuchElementException
+
+import com.fulcrumgenomics.cmdline.{ClpGroups, FgBioTool}
+import com.fulcrumgenomics.util.ReadStructure.SubRead
+import com.fulcrumgenomics.util._
+import com.fulcrumgenomics.util.miseq.{Sample, SampleBarcode, SampleSheet}
+import dagr.commons.CommonsDef.{DirPath, FilePath, PathToFastq, _}
+import dagr.commons.util.LazyLogging
+import dagr.sopt.{arg, clp}
+import htsjdk.samtools.SAMFileHeader.SortOrder
+import htsjdk.samtools._
+import htsjdk.samtools.fastq.FastqReader
+import htsjdk.samtools.util._
+
+import scala.collection.mutable
+
+object DemuxFastqs {
+  import SampleBarcode.SampleBarcodeDelimiter
+
+  val UnmatchedSampleId: String = "unmatched"
+
+  /** Custom columns in the Illumina Experiment Manager Sample Sheet. */
+  object CustomSampleSheetHeaderNames {
+    val R1SampleBarcodeBases = "R1_Barcode_Bases"
+    val R2SampleBarcodeBases = "R2_Barcode_Bases"
+  }
+
+  /** Creates the sample output BAM for the given sample.  [setSampleBarcode] should be called first. */
+  def sampleOutputBam(output: Path, sample: Sample): Path = {
+    val sampleBarcode: String = sample.sampleBarcode.getOrElse(throw new IllegalArgumentException("Sample barcode missing for sample: " + sample.sampleName)).toString
+    output.resolve(IOUtil.makeFileNameSafe(String.format("%s.%s.%s.bam", sample.sampleId, sample.sampleName, sampleBarcode)))
+  }
+
+  /** Gets the sample sheet after adding in any custom sample information tracked by DemuxFastqs. */
+  def sampleSheet(sampleSheet: Path, lane: Option[Int] = None): SampleSheet = {
+    new SampleSheet(
+      SampleSheet(sampleSheet, lane=lane)
+        .map(new CustomSample(_))
+        .toSeq
+    )
+  }
+
+  private val solexaQualityConverter: SolexaQualityConverter = SolexaQualityConverter.getSingleton
+
+  /** A custom sample class to support inline sample barcodes for read one and read two. */
+  private[fastq] class CustomSample(sample: Sample)
```

Comment:
It feels like there ought to be a way to do this with a simple trait that you could mix in at construction time, and maybe override how SampleSheet constructs it's Sample instances?